File: sdks/php/src/Model/Approval.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Model;

/**
 * A multi-signature approval document.
 */
final class Approval
{
    /**
     * @param array<string, string> $signatures
     */
    public function __construct(
        public readonly string $id,
        public readonly string $state = '',
        public readonly string $createdAt = '',
        public readonly ?string $validUntil = null,
        public readonly string $fromDid = '',
        public readonly string $toDid = '',
        public readonly string $via = '',
        public readonly string $parent = '',
        public readonly ?ApprovalSubject $subject = null,
        public readonly string $templateUrl = '',
        public readonly array $signatures = [],
        public readonly ?string $respondedAt = null,
        public readonly ?int $fanOut = null,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $subject = null;
        if (isset($data['subject']) && is_array($data['subject'])) {
            $subject = ApprovalSubject::fromArray($data['subject']);
        }

        return new self(
            id: (string) ($data['id'] ?? ''),
            state: (string) ($data['state'] ?? ''),
            createdAt: (string) ($data['created_at'] ?? ''),
            validUntil: $data['valid_until'] ?? null,
            fromDid: (string) ($data['from'] ?? ''),
            toDid: (string) ($data['to'] ?? ''),
            via: (string) ($data['via'] ?? ''),
            parent: (string) ($data['parent'] ?? ''),
            subject: $subject,
            templateUrl: (string) ($data['template_url'] ?? ''),
            signatures: $data['signatures'] ?? [],
            respondedAt: $data['responded_at'] ?? null,
            fanOut: $data['fan_out'] ?? null,
        );
    }
}

File: sdks/java/src/main/java/dev/atap/sdk/model/Approval.java
package dev.atap.sdk.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Map;

/**
 * A multi-signature approval document.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class Approval {

    @JsonProperty("id")
    private String id;

    @JsonProperty("state")
    private String state;

    @JsonProperty("created_at")
    private String createdAt;

    @JsonProperty("valid_until")
    private String validUntil;

    @JsonProperty("from")
    private String fromDid;

    @JsonProperty("to")
    private String toDid;

    @JsonProperty("via")
    private String via;

    @JsonProperty("parent")
    private String parent;

    @JsonProperty("subject")
    private ApprovalSubject subject;

    @JsonProperty("template_url")
    private String templateUrl;

    @JsonProperty("signatures")
    private Map<String, String> signatures;

    @JsonProperty("responded_at")
    private String respondedAt;

    @JsonProperty("fan_out")
    private Integer fanOut;

    public Approval() {
    }

    public String getId() { return id; }
    public void setId(String id) { this.id = id; }

    public String getState() { return state; }
    public void setState(String state) { this.state = state; }

    public String getCreatedAt() { return createdAt; }
    public void setCreatedAt(String createdAt) { this.createdAt = createdAt; }

    public String getValidUntil() { return validUntil; }
    public void setValidUntil(String validUntil) { this.validUntil = validUntil; }

    public String getFromDid() { return fromDid; }
    public void setFromDid(String fromDid) { this.fromDid = fromDid; }

    public String getToDid() { return toDid; }
    public void setToDid(String toDid) { this.toDid = toDid; }

    public String getVia() { return via; }
    public void setVia(String via) { this.via = via; }

    public String getParent() { return parent; }
    public void setParent(String parent) { this.parent = parent; }

    public ApprovalSubject getSubject() { return subject; }
    public void setSubject(ApprovalSubject subject) { this.subject = subject; }

    public String getTemplateUrl() { return templateUrl; }
    public void setTemplateUrl(String templateUrl) { this.templateUrl = templateUrl; }

    public Map<String, String> getSignatures() { return signatures; }
    public void setSignatures(Map<String, String> signatures) { this.signatures = signatures; }

    public String getRespondedAt() { return respondedAt; }
    public void setRespondedAt(String respondedAt) { this.respondedAt = respondedAt; }

    public Integer getFanOut() { return fanOut; }
    public void setFanOut(Integer fanOut) { this.fanOut = fanOut; }
}
